roaring/roaring.go

```diff
@@ -705,93 +674,98 @@ func (b *Bitmap) unionInPlace(others ...*Bitmap) {
-			// otherwise we have to allocate a new one or convert the existing container
-			// into a bitmap container.
-			if container == nil {
-				buf := make([]uint64, bitmapN)
-				ob := buf[:bitmapN]
-				container = &Container{
-					bitmap:        ob,
-					n:             0,
-					containerType: containerBitmap,
+			expectedN += summaryStats.n
+			var itersToUnion handledIters
+			if tContainer == nil {
+				if summaryStats.c == 1 {
+					// There's no target and we have only one container, we
+					// can just clone it instead of unioning
+					statsHit("unionInPlace/reuse")
+					target.Containers.Put(iKey, iContainer.Clone())
+					bitmapIters[i].handled = true
+					continue
+				}
+				// If we expect enough items that we'll want to convert to bitmap,
+				// and the container we already found isn't a bitmap, make a bitmap
+				// and union everything into that. Otherwise clone the first
+				// container and union everything else into it.
+				if expectedN >= 512 && iContainer.containerType != containerBitmap {
+					statsHit("unionInPlace/newBitmap")
+					tContainer = &Container{containerType: containerBitmap, bitmap: make([]uint64, bitmapN)}
+					itersToUnion = bitmapIters[i:]
+				} else {
+					statsHit("unionInPlace/clone")
+					tContainer = iContainer.Clone()
+					itersToUnion = bitmapIters[i+1:]
+				}
+			} else {
+				itersToUnion = bitmapIters[i:]
+				if expectedN >= 512 && tContainer.containerType != containerBitmap {
+					statsHit("unionInPlace/convertToBitmap")
+					switch tContainer.containerType {
+					case containerArray:
+						tContainer.arrayToBitmap()
+					case containerRun:
+						tContainer.runToBitmap()
+					}
 				}
-			} else if container.isArray() {
-				container.arrayToBitmap()
-			} else if container.isRun() {
-				container.runToBitmap()
 			}
 
-			// Once we've acquired a bitmap container (either by reusing the existing one
-			// or allocating a new one) then the last step is to iterate through all the
-			// other containers to see which ones have the same key, and union all of them
-			// into the target bitmap container. Only need to loop starting from i because
-			// anything previous to that has already been handled.
-			for j := i; j < len(bitmapIters); j++ {
-				jIter := bitmapIters[j]
-				jKey, jContainer := jIter.iter.Value()
+			// Now we union everything together. In many cases, this will mean
+			// converting a non-bitmap container to a bitmap, and then unioning things
+			// in, but we can specialize away from that in some cases.
+			//
+			// We iterate through the other Containers [plural] checking to see whether
+			// any of them have a Container [singular] with the same key to merge
+			// in. We only need to check i and higher, because anything lower would
+			// already be handled.
+			//
+			// If the target didn't previously have this index, we cloned the
+			// first container we found for it, so we will not then union that one
+			// in.
```

Comment:
"so we will not then union that one in since we've already set `itersToUnion` to skip it